public/app/features/explore/state/query.ts

```diff
@@ -486,11 +485,11 @@ async function handleHistory(
   queries: DataQuery[],
   exploreId: string
 ) {
-  const datasourceId = datasource.meta.id;
-  const nextHistory = updateHistory(history, datasourceId, queries);
-  dispatch(historyUpdatedAction({ exploreId, history: nextHistory }));
-
-  dispatch(addHistoryItem(datasource.uid, datasource.name, queries));
+  dispatch(addHistoryItem(true, datasource.uid, datasource.name, queries)); // always write to local for autocomplete, query history will use also if flag disabled
```

Comment:
Could it happen that:
- saving to local storage in line 488 fails (e.g. local storage is full)
- but then remote storage passes okay
- and the user see the error message for the local storage?
🤔 

It seems that for autocomplete we should ignore all the errors.

Reply:
Okay, I implemented this with another parameter in `addHistoryItem` that will mute all errors and warnings. This does make the method signature to `addToRichHistory` a bit ridiculous, as there are are already booleans for showing the quota exceeded error and the limit exceeded warning. But, there is a generic dispatch for all errors that pop up (hiding the duplicate error) and I thought it made sense to make that a separate "permission" of sorts. Let me know if you disagree.

Reply:
> This does make the method signature to `addToRichHistory` a bit ridiculous,

Would creating and using a wrapper for the remote storage make it more readable? 🤔  Something like:

```typescript
export default class DualWriteRichHistoryRemoteStorage extends RichHistoryRemoteStorage {

  autocompleteHistory: RichHistoryLocalStorage;
  
  constructor(props) {
    super(props);
    this.autocompleteHistory = new RichHistoryLocalStorage();
  }

  async addToRichHistory(
    newRichHistoryQuery: Omit<RichHistoryQuery, 'id' | 'createdAt'>
  ): Promise<{ warning?: RichHistoryStorageWarningDetails; richHistoryQuery: RichHistoryQuery }> {
    const result = await super.addToRichHistory(newRichHistoryQuery);
    try {
      this.autocompleteHistory.addToRichHistory(newRichHistoryQuery);
    } catch(e) {}
    return result;
  }

}
```

and then create DualWriteRichHistoryRemoteStorage when flag is enabled:

```typescript
const richHistoryLocalStorage = new RichHistoryLocalStorage();
const richHistoryRemoteStorage = new DualWriteRichHistoryRemoteStorage();

export const getRichHistoryStorage = (): RichHistoryStorage => {
  return config.queryHistoryEnabled ? richHistoryRemoteStorage : richHistoryLocalStorage;
};
```

Reply:
@ifrost the wrapper idea is a good one, but I just think another layer in all this makes it hard to follow. Can you check out what I have now? I made `richHistory/addToRichHistory` take an object with some sensible defaults, and I think we can make calling it more clean now without as much wrapper-ing.

Reply:
Looks good at first glance 👍  I'll try to test it locally tomorrow. Do you think renaming actions a bit would make sense to reflect what they mean now and have:

```
richHistoryStorageFullErrorShownAction
richHistoryLimitExceededWarningShownAction
```
?